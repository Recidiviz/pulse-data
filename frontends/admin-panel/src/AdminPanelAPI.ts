// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2020 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================
const postWithURLAndBody = async (
  url: string,
  body: Record<string, unknown>
): Promise<Response> => {
  return fetch(`/admin${url}`, {
    method: "POST",
    body: JSON.stringify(body),
    headers: {
      "Content-Type": "application/json",
    },
  });
};

// Fetch state dataset info
export const fetchColumnObjectCountsByValue = async (
  table: string,
  column: string
): Promise<Response> => {
  return postWithURLAndBody(
    "/api/ingest_metadata/fetch_column_object_counts_by_value",
    { table, column }
  );
};

export const fetchTableNonNullCountsByColumn = async (
  table: string
): Promise<Response> => {
  return postWithURLAndBody(
    "/api/ingest_metadata/fetch_table_nonnull_counts_by_column",
    { table }
  );
};

export const fetchObjectCountsByTable = async (): Promise<Response> => {
  return postWithURLAndBody(
    "/api/ingest_metadata/fetch_object_counts_by_table",
    {}
  );
};

// Fetch data freshness
export const fetchDataFreshness = async (): Promise<Response> => {
  return postWithURLAndBody("/api/ingest_metadata/data_freshness", {});
};
