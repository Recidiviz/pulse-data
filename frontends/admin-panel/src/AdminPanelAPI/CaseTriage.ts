// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2021 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================
import { postWithURLAndBody } from "./utils";

// Cloud SQL -> GCS CSV Export
export const generateCaseUpdatesExport = async (): Promise<Response> => {
  return postWithURLAndBody("/api/case_triage/generate_case_updates_export");
};

// GCS CSV -> Cloud SQL Import
export const fetchETLViewIds = async (): Promise<Response> => {
  return postWithURLAndBody("/api/case_triage/fetch_etl_view_ids");
};

export const runCloudSQLImport = async (
  viewIds: string[]
): Promise<Response> => {
  return postWithURLAndBody("/api/case_triage/run_gcs_import", {
    viewIds,
  });
};

// PO Feedback
export const getPOFeedback = async (): Promise<Response> => {
  return postWithURLAndBody("/api/case_triage/get_po_feedback");
};

// Fetch states for po monthly reports
export const fetchEmailStateCodes = async (): Promise<Response> => {
  return postWithURLAndBody(
    // change to fetch_email_state_codes
    "/api/ingest_operations/fetch_ingest_state_codes"
  );
};
