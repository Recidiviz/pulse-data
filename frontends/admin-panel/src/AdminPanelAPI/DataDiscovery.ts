// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2021 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================
import { postWithURLAndBody } from "./utils";

export const fetchRegionCodeFiles = async (
  regionCode: string
): Promise<Response> => {
  return postWithURLAndBody("/data_discovery/files", {
    region_code: regionCode,
  });
};

export interface Message {
  cursor: number;
  data: string;
  kind: string;
}

export const fetchDiscoveryStatus = async (
  discoveryId: string,
  messageCursor: number
): Promise<Response> => {
  return postWithURLAndBody("/data_discovery/discovery_status", {
    message_cursor: messageCursor,
    discovery_id: discoveryId,
  });
};

export const pollDiscoveryStatus = async (
  discoveryId: string,
  latestMessage: Message | null,
  onMessageReceived: CallableFunction
): Promise<void> => {
  const response = await fetchDiscoveryStatus(
    discoveryId,
    latestMessage ? latestMessage.cursor : 0
  );

  if (response.status === 200) {
    // Get and show the message
    const receivedMessage = await response.json();

    onMessageReceived(receivedMessage);

    if (receivedMessage.kind === "close") {
      return;
    }

    await pollDiscoveryStatus(discoveryId, receivedMessage, onMessageReceived);
  }

  if (response.status === 500) {
    onMessageReceived(null);
  }
};

export const createDiscovery = async (
  body: Record<string, unknown>
): Promise<Response> => {
  return postWithURLAndBody("/data_discovery/create_discovery", body);
};
