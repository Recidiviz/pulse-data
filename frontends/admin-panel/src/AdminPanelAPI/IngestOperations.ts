// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2021 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================
import {
  DirectIngestInstance,
  QueueState,
} from "../components/IngestOperationsView/constants";
import { getResource, postWithURLAndBody } from "./utils";

// Fetch states with ingest
export const fetchIngestStateCodes = async (): Promise<Response> => {
  return postWithURLAndBody("/api/ingest_operations/fetch_ingest_state_codes");
};

//  Trigger task scheduler
export const triggerTaskScheduler = async (
  regionCode: string,
  instance: DirectIngestInstance
): Promise<Response> => {
  return postWithURLAndBody(
    `/api/ingest_operations/${regionCode}/trigger_task_scheduler`,
    { instance }
  );
};

// Update ingest queue states
export const updateIngestQueuesState = async (
  regionCode: string,
  newQueueState: QueueState
): Promise<Response> => {
  return postWithURLAndBody(
    `/api/ingest_operations/${regionCode}/update_ingest_queues_state`,
    { new_queue_state: newQueueState }
  );
};

// Get ingest queue states
export const getIngestQueuesState = async (
  regionCode: string
): Promise<Response> => {
  return fetch(
    `/admin/api/ingest_operations/${regionCode}/get_ingest_queue_states`,
    {
      headers: {
        "Content-Type": "application/json",
      },
    }
  );
};

// Get ingest instance summary
export const getIngestInstanceSummary = async (
  regionCode: string,
  ingestInstance: DirectIngestInstance
): Promise<Response> => {
  return fetch(
    `/admin/api/ingest_operations/${regionCode}/get_ingest_instance_summary/${ingestInstance}`,
    {
      headers: {
        "Content-Type": "application/json",
      },
    }
  );
};

// Get ingest raw file processing status
export const getIngestRawFileProcessingStatus = async (
  regionCode: string,
  ingestInstance: DirectIngestInstance
): Promise<Response> => {
  return fetch(
    `/admin/api/ingest_operations/${regionCode}/get_ingest_raw_file_processing_status/${ingestInstance}`,
    {
      headers: {
        "Content-Type": "application/json",
      },
    }
  );
};

// Start CloudSQL export to GCS
export const exportDatabaseToGCS = async (
  stateCode: string,
  ingestInstance: DirectIngestInstance
): Promise<Response> => {
  return postWithURLAndBody(`/api/ingest_operations/export_database_to_gcs`, {
    stateCode,
    ingestInstance,
  });
};

// Start CloudSQL import from GCS
export const importDatabaseFromGCS = async (
  stateCode: string,
  importToDatabaseInstance: DirectIngestInstance,
  exportedDatabaseInstance: DirectIngestInstance
): Promise<Response> => {
  return postWithURLAndBody(`/api/ingest_operations/import_database_from_gcs`, {
    stateCode,
    importToDatabaseInstance,
    exportedDatabaseInstance,
  });
};

// Clean up CloudSQL import files
export const deleteDatabaseImportGCSFiles = async (
  stateCode: string,
  exportedDatabaseInstance: DirectIngestInstance
): Promise<Response> => {
  return postWithURLAndBody(
    `/api/ingest_operations/delete_database_import_gcs_files`,
    {
      stateCode,
      exportedDatabaseInstance,
    }
  );
};

// Acquire BQ Export lock for the STATE database
export const acquireBQExportLock = async (
  stateCode: string,
  ingestInstance: DirectIngestInstance
): Promise<Response> => {
  return postWithURLAndBody(`/api/ingest_operations/acquire_ingest_lock`, {
    stateCode,
    ingestInstance,
  });
};

// Release BQ Export lock for the STATE database
export const releaseBQExportLock = async (
  stateCode: string,
  ingestInstance: DirectIngestInstance
): Promise<Response> => {
  return postWithURLAndBody(`/api/ingest_operations/release_ingest_lock`, {
    stateCode,
    ingestInstance,
  });
};

// Pauses direct ingest instance
export const pauseDirectIngestInstance = async (
  stateCode: string,
  ingestInstance: DirectIngestInstance
): Promise<Response> => {
  return postWithURLAndBody(
    `/api/ingest_operations/pause_direct_ingest_instance`,
    {
      stateCode,
      ingestInstance,
    }
  );
};

// Unpauses direct ingest instance
export const unpauseDirectIngestInstance = async (
  stateCode: string,
  ingestInstance: DirectIngestInstance
): Promise<Response> => {
  return postWithURLAndBody(
    `/api/ingest_operations/unpause_direct_ingest_instance`,
    {
      stateCode,
      ingestInstance,
    }
  );
};

// Import raw files to BiqQuery Sandbox
export const importRawDataToSandbox = async (
  stateCode: string,
  sandboxDatasetPrefix: string,
  sourceBucket: string,
  fileTagFilters: string[] | undefined
): Promise<Response> => {
  return postWithURLAndBody(
    `/api/ingest_operations/direct/sandbox_raw_data_import`,
    {
      stateCode,
      sandboxDatasetPrefix,
      sourceBucket,
      fileTagFilters,
    }
  );
};

// Get list of sandbox buckets
export const listSandboxBuckets = async (): Promise<Response> => {
  return postWithURLAndBody(
    `/api/ingest_operations/direct/list_sandbox_buckets`
  );
};

// Get list of raw files and dates in sandbox bucket
export const listRawFilesInSandboxBucket = async (
  stateCode: string,
  sourceBucket: string
): Promise<Response> => {
  return postWithURLAndBody(`/api/ingest_operations/direct/list_raw_files`, {
    stateCode,
    sourceBucket,
  });
};

// Move ingest view results to backup dataset
export const moveIngestViewResultsToBackup = async (
  stateCode: string,
  ingestInstance: DirectIngestInstance
): Promise<Response> => {
  return postWithURLAndBody(
    "/api/ingest_operations/flash_primary_db/move_ingest_view_results_to_backup",
    {
      stateCode,
      ingestInstance,
    }
  );
};

// Move ingest view results between instances
export const moveIngestViewResultsBetweenInstances = async (
  stateCode: string,
  srcIngestInstance: DirectIngestInstance,
  destIngestInstance: DirectIngestInstance
): Promise<Response> => {
  return postWithURLAndBody(
    "/api/ingest_operations/flash_primary_db/move_ingest_view_results_between_instances",
    {
      stateCode,
      srcIngestInstance,
      destIngestInstance,
    }
  );
};

// Mark instance ingest view data as invalidated
export const markInstanceIngestViewDataInvalidated = async (
  stateCode: string,
  ingestInstance: DirectIngestInstance
): Promise<Response> => {
  return postWithURLAndBody(
    "/api/ingest_operations/flash_primary_db/mark_instance_ingest_view_data_invalidated",
    {
      stateCode,
      ingestInstance,
    }
  );
};

// Transer ingest view metadata to new instance
export const transferIngestViewMetadataToNewInstance = async (
  stateCode: string,
  srcIngestInstance: DirectIngestInstance,
  destIngestInstance: DirectIngestInstance
): Promise<Response> => {
  return postWithURLAndBody(
    "/api/ingest_operations/flash_primary_db/transfer_ingest_view_metadata_to_new_instance",
    {
      stateCode,
      srcIngestInstance,
      destIngestInstance,
    }
  );
};

// Get all ingest instance statuses
export const getAllIngestInstanceStatuses = async (): Promise<Response> => {
  return getResource("/api/ingest_operations/all_ingest_instance_statuses");
};
