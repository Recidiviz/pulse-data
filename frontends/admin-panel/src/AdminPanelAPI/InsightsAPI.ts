// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================
import { StateCodeInfo } from "../components/general/constants";
import {
  InsightsConfiguration,
  insightsConfigurationSchema,
} from "../InsightsStore/models/InsightsConfiguration";
import { get, post, put } from "./utils";

// Get enabled states for Insights
export const getInsightsStateCodeInfo = async (): Promise<StateCodeInfo[]> => {
  return get("/admin/outliers/enabled_state_codes");
};

export const getInsightsConfigs = async (
  stateCode: string
): Promise<InsightsConfiguration[]> => {
  const response = await get(`/admin/outliers/${stateCode}/configurations`);
  const configs = response as Array<unknown>;
  return configs.map((c) => {
    return insightsConfigurationSchema.parse(c);
  });
};

export const createNewConfig = async (
  request: InsightsConfiguration,
  stateCode: string
): Promise<Response> => {
  return post(`/admin/outliers/${stateCode}/configurations`, request);
};

export const promoteConfigToProduction = async (
  configId: number,
  stateCode: string
): Promise<Response> => {
  return post(
    `/admin/outliers/${stateCode}/configurations/${configId}/promote/production`
  );
};

export const promoteConfigToDefault = async (
  configId: number,
  stateCode: string
): Promise<Response> => {
  return post(
    `/admin/outliers/${stateCode}/configurations/${configId}/promote/default`
  );
};

export const reactivateConfig = async (
  configId: number,
  stateCode: string
): Promise<Response> => {
  return post(
    `/admin/outliers/${stateCode}/configurations/${configId}/reactivate`
  );
};

export const deactivateConfig = async (
  configId: number,
  stateCode: string
): Promise<Response> => {
  return put(
    `/admin/outliers/${stateCode}/configurations/${configId}/deactivate`
  );
};
