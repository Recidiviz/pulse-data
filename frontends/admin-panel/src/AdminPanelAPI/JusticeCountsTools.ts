// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2021 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { User } from "../components/JusticeCountsTools/constants";
import { getResource, postWithURLAndBody, putWithURLAndBody } from "./utils";

// Agency Provisioning
export const getAgencies = async (): Promise<Response> => {
  return getResource(`/api/justice_counts_tools/agencies`);
};

export const createAgency = async (
  name: string,
  systems: string[],
  stateCode: string,
  fipsCountyCode?: string
): Promise<Response> => {
  return postWithURLAndBody(`/api/justice_counts_tools/agencies`, {
    name,
    systems,
    state_code: stateCode,
    fips_county_code: fipsCountyCode,
  });
};

// User Provisioning
export const getUsers = async (): Promise<Response> => {
  return getResource(`/api/justice_counts_tools/users`);
};

export const updateUser = async (
  user: User,
  name: string | null,
  agencyIds: number[] | null
): Promise<Response> => {
  return putWithURLAndBody(`/api/justice_counts_tools/users`, {
    name,
    auth0_user_id: user.auth0_user_id,
    agency_ids: agencyIds,
  });
};

export const bulkUpload = async (formData: FormData): Promise<Response> => {
  // Don't use postWithURLAndBody because that will JSONify payload,
  // and we don't want to do that because it contains a file
  const url = `/api/justice_counts_tools/bulk_upload`;
  return fetch(`/admin${url}`, {
    method: "POST",
    body: formData,
  });
};
