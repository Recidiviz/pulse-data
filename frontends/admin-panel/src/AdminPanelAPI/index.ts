// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2020 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import MetadataDataset from "../models/MetadataDatasets";
import { postWithURLAndBody } from "./utils";

// Fetch dataset metadata
export const fetchColumnObjectCountsByValue = async (
  metadataDataset: MetadataDataset,
  table: string,
  column: string
): Promise<Response> => {
  return postWithURLAndBody(
    `/api/${metadataDataset}/fetch_column_object_counts_by_value`,
    { table, column }
  );
};

export const fetchTableNonNullCountsByColumn = async (
  metadataDataset: MetadataDataset,
  table: string
): Promise<Response> => {
  return postWithURLAndBody(
    `/api/${metadataDataset}/fetch_table_nonnull_counts_by_column`,
    { table }
  );
};

export const fetchObjectCountsByTable = async (
  metadataDataset: MetadataDataset
): Promise<Response> => {
  return postWithURLAndBody(
    `/api/${metadataDataset}/fetch_object_counts_by_table`
  );
};

// Fetch data freshness
export const fetchDataFreshness = async (): Promise<Response> => {
  return postWithURLAndBody("/api/ingest_metadata/data_freshness");
};

// Fetch validation status
export const fetchValidationStateCodes = async (): Promise<Response> => {
  return postWithURLAndBody("/api/validation_metadata/state_codes");
};

export const fetchValidationStatus = async (): Promise<Response> => {
  return postWithURLAndBody("/api/validation_metadata/status");
};

export const fetchValidationDetails = async (
  validationName: string,
  stateCode: string
): Promise<Response> => {
  return postWithURLAndBody(
    `/api/validation_metadata/status/${validationName}/${stateCode}`
  );
};

export const fetchValidationDescription = async (
  validationName: string
): Promise<Response> => {
  return postWithURLAndBody(
    `/api/validation_metadata/description/${validationName}`
  );
};

export const fetchValidationErrorTable = async (
  validationName: string,
  stateCode: string
): Promise<Response> => {
  return postWithURLAndBody(
    `/api/validation_metadata/error_table/${validationName}/${stateCode}`
  );
};

export {
  acquireBQExportLock,
  deleteDatabaseImportGCSFiles,
  exportDatabaseToGCS,
  fetchIngestStateCodes,
  getIngestInstanceSummary,
  getIngestFileProcessingStatus,
  getIngestQueuesState,
  importDatabaseFromGCS,
  markInstanceIngestViewDataInvalidated,
  moveIngestViewResultsBetweenInstances,
  moveIngestViewResultsToBackup,
  pauseDirectIngestInstance,
  releaseBQExportLock,
  triggerTaskScheduler,
  transferIngestViewMetadataToNewInstance,
  unpauseDirectIngestInstance,
  updateIngestQueuesState,
} from "./IngestOperations";
export { createAgency, getAgencies, getUsers } from "./JusticeCountsTools";
export {
  fetchEmailStateCodes,
  fetchETLViewIds,
  fetchRawFilesStateCodes,
  fetchRosterStateCodes,
  generateNonETLExports,
  getPOFeedback,
  runCloudSQLImport,
  getStateUserPermissions,
} from "./LineStaffTools";
