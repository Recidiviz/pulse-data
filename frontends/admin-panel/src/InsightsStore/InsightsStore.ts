// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { action, autorun, flowResult, makeAutoObservable } from "mobx";

import {
  getInsightsConfigurations,
  getInsightsStateCodeInfo,
} from "../AdminPanelAPI/InsightsAPI";
import { StateCodeInfo } from "../components/general/constants";
import { InsightsConfiguration } from "./models/InsightsConfiguration";
import ConfigurationPresenter from "./presenters/ConfigurationPresenter";
import { Hydratable, HydrationState } from "./types";

export class InsightsStore implements Hydratable {
  hydrationState: HydrationState;

  stateCode?: string;

  stateCodeInfo?: StateCodeInfo[];

  configs?: InsightsConfiguration[];

  configurationPresenter?: ConfigurationPresenter;

  constructor() {
    makeAutoObservable(
      this,
      { initializeConfigurationPresenter: action },
      { autoBind: true }
    );
    this.hydrationState = { status: "needs hydration" };

    autorun(() => {
      if (this.stateCode) {
        this.initializeConfigurationPresenter();
      }
    });
  }

  setStateCode(stateCode: string): void {
    this.resetStateData();
    this.stateCode = stateCode;
  }

  resetStateData() {
    this.configs = undefined;
    this.configurationPresenter = undefined;
  }

  initializeConfigurationPresenter() {
    if (!this.stateCode) return;
    this.configurationPresenter = new ConfigurationPresenter(
      this,
      this.stateCode
    );
  }

  *populateConfigs(): Generator<
    Promise<InsightsConfiguration[]>,
    void,
    InsightsConfiguration[]
  > {
    if (!this.stateCode) throw new Error("missing state code");
    this.configs = yield getInsightsConfigurations(this.stateCode);
  }

  *populateStateCodeInfo(): Generator<
    Promise<StateCodeInfo[]>,
    void,
    StateCodeInfo[]
  > {
    this.stateCodeInfo = yield getInsightsStateCodeInfo();
  }

  async hydrate(): Promise<void> {
    if (
      this.hydrationState.status === "hydrated" ||
      this.hydrationState.status === "loading"
    )
      return;

    try {
      this.hydrationState = { status: "loading" };
      await flowResult(this.populateStateCodeInfo());
      this.hydrationState = { status: "hydrated" };
    } catch (e) {
      this.hydrationState = { status: "failed", error: e as Error };
    }
  }
}
