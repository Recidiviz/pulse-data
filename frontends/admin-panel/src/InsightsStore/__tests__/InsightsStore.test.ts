// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { configure } from "mobx";

import { getInsightsStateCodeInfo } from "../../AdminPanelAPI/InsightsAPI";
import { InsightsStore } from "../InsightsStore";
import { rawStateCodeInfoFixture } from "../models/fixtures/StateCodeInfoFixture";

jest.mock("../../AdminPanelAPI/InsightsAPI");
const mockGetInsightsStateCodeInfo = getInsightsStateCodeInfo as jest.Mock;

let store: InsightsStore;

beforeEach(async () => {
  jest.resetModules();
  configure({ safeDescriptors: false });
  store = new InsightsStore();
  mockGetInsightsStateCodeInfo.mockResolvedValue(rawStateCodeInfoFixture);
});

afterEach(() => {
  jest.restoreAllMocks();
  configure({ safeDescriptors: true });
});

test("hydrate store", async () => {
  expect(store.hydrationState.status).toEqual("needs hydration");
  await store.hydrate();
  expect(store.hydrationState.status).toEqual("hydrated");
});

test("hydration -- error", async () => {
  const error = new Error("hydrationError");
  mockGetInsightsStateCodeInfo.mockRejectedValue(error);
  await store.hydrate();
  expect(store.hydrationState).toEqual({
    status: "failed",
    error,
  });
});

test("stateCodeInfo", async () => {
  await store.hydrate();

  expect(store.stateCodeInfo).toMatchSnapshot();
});
