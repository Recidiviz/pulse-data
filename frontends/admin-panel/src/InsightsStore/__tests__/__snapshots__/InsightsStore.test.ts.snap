// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stateCodeInfo 1`] = `
Array [
  Object {
    "code": "US_PA",
    "name": "Pennsylvania",
  },
  Object {
    "code": "US_IX",
    "name": "Idaho ATLAS",
  },
  Object {
    "code": "US_MI",
    "name": "Michigan",
  },
  Object {
    "code": "US_TN",
    "name": "Tennessee",
  },
]
`;
